packages/relay/src/services/ReportService.ts

```diff
@@ -280,6 +295,389 @@ export class ReportService {
+        reportId: string,
+        nullifier?: string | undefined,
+        repUserType?: RepUserType
+    ) {
+        if (type === ReputationDirection.POSITIVE && nullifier) {
+            const report = await this.findReportWithNullifier(
+                db,
+                reportId,
+                nullifier,
+                ReportStatus.WAITING_FOR_TRANSACTION
+            )
+
+            if (report) {
+                const updatedAdjudicators = report.adjudicatorsNullifier.map(
+                    (adj) =>
+                        adj.nullifier === nullifier
+                            ? { ...adj, claimed: true }
+                            : adj
+                )
+
+                await db.update('ReportHistory', {
+                    where: { _id: report._id },
+                    update: {
+                        adjudicatorsNullifier: updatedAdjudicators,
+                    },
+                })
+            }
+        } else {
+            let updateField: string
+
+            if (type === ReputationDirection.POSITIVE) {
+                updateField = 'reportorClaimedRep'
+            } else {
+                if (repUserType === RepUserType.REPORTER) {
+                    updateField = 'reportorClaimedRep'
+                } else if (repUserType === RepUserType.POSTER) {
+                    updateField = 'respondentClaimedRep'
+                } else {
+                    throw new Error(
+                        'Invalid repUserType for negative reputation'
+                    )
+                }
+            }
+
+            const updateQuery = { [updateField]: true }
+            await db.update('ReportHistory', {
+                where: {
+                    reportId,
+                },
+                update: updateQuery,
+            })
+        }
+    }
+
+    async findReportWithNullifier(
+        db: DB,
+        reportId: string,
+        nullifier: string,
+        status: ReportStatus
+    ) {
+        const reports = await db.findMany('ReportHistory', {
+            where: {
+                reportId,
+                status: status,
+            },
+        })
+
+        return reports.find((report) =>
+            report.adjudicatorsNullifier.some(
+                (adj) => adj.nullifier === nullifier
+            )
+        )
+    }
+
+    private getReputationType(
+        direction: ReputationDirection,
+        repUserType: RepUserType
+    ): ReputationType {
+        if (direction === ReputationDirection.POSITIVE) {
+            return repUserType === RepUserType.VOTER
+                ? ReputationType.ADJUDICATE
+                : ReputationType.REPORT_SUCCESS
+        } else if (direction === ReputationDirection.NEGATIVE) {
+            return repUserType === RepUserType.POSTER
+                ? ReputationType.BE_REPORTED
+                : ReputationType.REPORT_FAILURE
+        } else {
+            throw new Error('Invalid ReputationDirection')
+        }
+    }
```

Comment:
This implementation will return wrong value below these situations:
- `direction` is `ReputationDirection.NEGATIVE` and `repUserType` is `RepUserType.Voter`
- `direction` is `ReputationDirection.POSITIVE` and `repUserType` is `RepUserType.POSTER`

Reply:
fixed